bundledata.go

```diff
@@ -14,16 +15,62 @@ import (
 	"strconv"
 	"strings"
 
-	"github.com/juju/names"
-	"gopkg.in/yaml.v1"
+	"gopkg.in/juju/names.v2"
+	"gopkg.in/yaml.v2"
 )
 
+type bundleDataCompat struct {
+	// LegacyServices holds application entries for older bundle files
```

Comment:
How about just embedding BundleData into this struct?
Something like this is simpler, I think, and consolidates the legacy checking
into one place:

	type noMethodsBundleData BundleData
	type legacyBundleData struct {
		noMethodsBundleData `bson:",inline" yaml:",inline"`
		Services            map[string]*ApplicationSpec `json:"services" yaml:"services" bson:"services"`
	}
	
	func (bdc *legacyBundleData) setBundleData(bd *BundleData) error {
		if len(bdc.Applications) > 0 && len(bdc.Services) > 0 {
			return fmt.Errorf("cannot specify both applications and services")
		}
		if len(bdc.Services) > 0 {
			bdc.Applications = bdc.Services
		}
		*bd = BundleData(bdc.noMethodsBundleData)
		return nil
	}
	
	func (bd *BundleData) UnmarshalJSON(b []byte) error {
		var bdc legacyBundleData
		if err := json.Unmarshal(b, &bdc); err != nil {
			return err
		}
		return bdc.setBundleData(bd)
	}
	
	func (bd *BundleData) UnmarshalYAML(f func(interface{}) error) error {
		var bdc legacyBundleData
		if err := f(&bdc); err != nil {
			return err
		}
		return bdc.setBundleData(bd)
	}
	
	func (bd *BundleData) SetBSON(raw bson.Raw) error {
		var bdc legacyBundleData
		if err := raw.Unmarshal(&bdc); err != nil {
			return err
		}
		return bdc.setBundleData(bd)
	}